src/back/spirv/layout.rs

```diff
@@ -0,0 +1,216 @@
+    pub(crate) debugs: Vec<Word>,
+    pub(crate) annotations: Vec<Word>,
+    pub(crate) type_declarations: Vec<Word>,
+    pub(crate) constants: Vec<Word>,
+    pub(crate) global_variables: Vec<Word>,
+    function_declarations: Vec<Word>,
+    pub(crate) function_definitions: Vec<Word>,
+}
+
+impl LogicalLayout {
+    pub(crate) fn new() -> Self {
+        LogicalLayout {
+            capabilities: vec![],
+            extensions: vec![],
+            ext_inst_imports: vec![],
+            memory_model: vec![],
+            entry_points: vec![],
+            execution_modes: vec![],
+            debugs: vec![],
+            annotations: vec![],
+            type_declarations: vec![],
+            constants: vec![],
+            global_variables: vec![],
+            function_declarations: vec![],
+            function_definitions: vec![],
+        }
+    }
+
+    pub(crate) fn in_words(&self) -> Vec<u32> {
+        let mut words: Vec<Word> = vec![];
+
+        for capability in self.capabilities.iter() {
+            words.push(*capability);
+        }
+
+        for extension in self.extensions.iter() {
+            words.push(*extension);
+        }
+
+        for ext_inst_import in self.ext_inst_imports.iter() {
+            words.push(*ext_inst_import);
+        }
+
+        for memory_model in self.memory_model.iter() {
+            words.push(*memory_model);
+        }
+
+        for entry_point in self.entry_points.iter() {
+            words.push(*entry_point);
+        }
+
+        for execution_mode in self.execution_modes.iter() {
+            words.push(*execution_mode);
+        }
+
+        for debug in self.debugs.iter() {
+            words.push(*debug);
+        }
+
+        for annotation in self.annotations.iter() {
+            words.push(*annotation);
+        }
+
+        for type_declaration in self.type_declarations.iter() {
+            words.push(*type_declaration);
+        }
+
+        for constants in self.constants.iter() {
+            words.push(*constants);
+        }
+
+        for global_variables in self.global_variables.iter() {
+            words.push(*global_variables);
+        }
+
+        for function_declaration in self.function_declarations.iter() {
+            words.push(*function_declaration);
+        }
+
+        for function_definition in self.function_definitions.iter() {
+            words.push(*function_definition);
+        }
+
+        words
+    }
+}
+
+pub(crate) struct Instruction {
+    op: u32,
+    wc: u32,
+    type_id: Option<Word>,
+    result_id: Option<Word>,
+    operands: Vec<Word>,
+}
+
+impl Instruction {
+    pub(crate) fn new(op: Op) -> Self {
+        Instruction {
+            op: op as u32,
+            wc: 1, // Always start at 1 for the first word (OP + WC),
+            type_id: None,
+            result_id: None,
+            operands: vec![],
+        }
+    }
+
+    pub(crate) fn set_type(&mut self, id: Word) {
+        if self.type_id.is_none() {
```

Comment:
Can we instead assert that the type id is None? I don't think we expect ourselves to call this multiple times on the same instruction.

Reply:
Yes that is true, I can do something like: 
```rust
assert!(self.type_id.is_some(), "Type can only be set once");
self.type_id = Some(id);
self.wc += 1;
```